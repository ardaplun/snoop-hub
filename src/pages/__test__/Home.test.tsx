import React from 'react';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import OnTestedComponent from '../../router';

beforeAll(() => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
  const noop = () => {};
  Object.defineProperty(window, 'scrollTo', { value: noop, writable: true });
});

describe('<MainPage />', () => {
  let component = <OnTestedComponent />

  it('properly show loader', async () => {
    const { asFragment, findByTestId } = render(component);
    const loader = await findByTestId('loader');

    expect(asFragment()).toMatchSnapshot();
    expect(loader).toBeInTheDocument();
  });

  it('properly show main page', async () => {
    const { asFragment, findByTestId, queryByText } = render(component);
    const inputSearch = await findByTestId('input-search');
    const welcomeMessage = queryByText(/Welcome to Snoop-hub/i);

    expect(welcomeMessage).toBeInTheDocument();
    expect(inputSearch).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('input properly do receive value', async () => {
    const { asFragment, getByPlaceholderText } = render(component);
    const inputSearch = getByPlaceholderText(/Search github user/i);
    fireEvent.change(inputSearch, { target: { value: 'ardaplun' } });

    expect(inputSearch.value).toBe('ardaplun');
    expect(asFragment()).toMatchSnapshot();
  });

  it('search by inputted value', async () => {
    const { asFragment, queryByText, queryAllByText, getByPlaceholderText } = render(component);
    const inputSearch = getByPlaceholderText(/Search github user/i);
    
    fireEvent.change(inputSearch, { target: { value: 'ardaplun' } });
    fireEvent.keyDown(inputSearch, { key: 'Enter', keyCode: 13 });
    
    expect(inputSearch).toHaveProperty('disabled');

    await waitForElement(() => queryByText(/Found/i));
    const foundText = queryByText(/Found/i);
    const userNameFound = queryAllByText(/ardaplun/i);

    expect(foundText).toBeInTheDocument();
    expect(userNameFound).toHaveLength(2);
    expect(asFragment()).toMatchSnapshot();
  });

  it('click on search result redirect to user page', async () => {

    const { asFragment, getByPlaceholderText, findByTestId, queryByText, queryAllByText } = render(component);
    const inputSearch = getByPlaceholderText(/Search github user/i);

    fireEvent.change(inputSearch, { target: { value: 'ardaplun' } });
    fireEvent.keyDown(inputSearch, { key: 'Enter', keyCode: 13 });

    await waitForElement(() => queryByText(/Found/i));
    const userNameFound = queryAllByText(/ardaplun/i);
    fireEvent.click(userNameFound[1]);

    // check loader on change page
    await waitForElement(() => findByTestId('loader'));
    const loader = await findByTestId('loader');
    expect(loader).toBeInTheDocument();

    await waitForElement(() => queryByText(/Welcome/i));

    expect(asFragment()).toMatchSnapshot();
  });

})