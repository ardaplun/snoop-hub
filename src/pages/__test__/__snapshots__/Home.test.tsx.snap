// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainPage /> click on search result redirect to user page 1`] = `
<DocumentFragment>
  <main
    class="container"
  >
    <div
      class="ant-page-header ant-page-header-ghost"
      style="border-bottom: 1px solid #ccc; width: 100%;"
    >
      <div
        class="ant-page-header-heading"
      >
        <div
          class="ant-page-header-heading-left"
        >
          <div
            class="ant-page-header-back"
          >
            <div
              aria-label="Back"
              class="ant-page-header-back-button"
              role="button"
              style="border: 0px; background: transparent; padding: 0px; line-height: inherit; display: inline-block;"
              tabindex="0"
            >
              <span
                aria-label="arrow-left"
                class="anticon anticon-arrow-left"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="arrow-left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M872 474H286.9l350.2-304c5.6-4.9 2.2-14-5.2-14h-88.5c-3.9 0-7.6 1.4-10.5 3.9L155 487.8a31.96 31.96 0 000 48.3L535.1 866c1.5 1.3 3.3 2 5.2 2h91.5c7.4 0 10.8-9.2 5.2-14L286.9 550H872c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <span
            class="ant-page-header-heading-title"
            title="Welcome to ardaplun page"
          >
            Welcome to ardaplun page
          </span>
        </div>
      </div>
    </div>
    <div
      class="loader "
      data-testid="loader"
    >
      <div
        class="ant-spin ant-spin-lg ant-spin-spinning"
      >
        <span
          class="ant-spin-dot ant-spin-dot-spin"
        >
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
        </span>
      </div>
    </div>
    <h4
      class="ant-typography"
    />
    <div
      class="Toastify"
    />
  </main>
</DocumentFragment>
`;

exports[`<MainPage /> input properly do receive value 1`] = `
<DocumentFragment>
  <main
    class="container"
  >
    <div
      style="margin: 100px 0px 16px; padding: 0px 16px; width: 100%;"
    >
      <span
        class="ant-input-search ant-input-search-enter-button ant-input-group-wrapper"
      >
        <span
          class="ant-input-wrapper ant-input-group"
        >
          <input
            class="ant-input"
            data-testid="input-search"
            placeholder="Search github user"
            type="text"
            value="ardaplun"
          />
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn ant-input-search-button ant-btn-primary"
              type="button"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
      </span>
    </div>
    <div
      style="text-align: center;"
    />
    <div
      style="flex: 1; width: 100%; padding: 16px; overflow-y: auto;"
    >
      <div
        style="text-align: center;"
      >
        <h4
          class="ant-typography"
          data-testid="text-welcome"
        >
          Welcome to Snoop-hub :)
        </h4>
      </div>
      <div
        style="display: flex; justify-content: center; margin: 16px 0px 40px;"
      />
    </div>
    <div
      class="Toastify"
    />
  </main>
</DocumentFragment>
`;

exports[`<MainPage /> properly show loader 1`] = `
<DocumentFragment>
  <div
    class="loader fullPage"
    data-testid="loader"
  >
    <div
      class="ant-spin ant-spin-lg ant-spin-spinning"
    >
      <span
        class="ant-spin-dot ant-spin-dot-spin"
      >
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<MainPage /> properly show main page 1`] = `
<DocumentFragment>
  <main
    class="container"
  >
    <div
      style="margin: 100px 0px 16px; padding: 0px 16px; width: 100%;"
    >
      <span
        class="ant-input-search ant-input-search-enter-button ant-input-group-wrapper"
      >
        <span
          class="ant-input-wrapper ant-input-group"
        >
          <input
            class="ant-input"
            data-testid="input-search"
            placeholder="Search github user"
            type="text"
            value=""
          />
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn ant-input-search-button ant-btn-primary"
              type="button"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
      </span>
    </div>
    <div
      style="text-align: center;"
    />
    <div
      style="flex: 1; width: 100%; padding: 16px; overflow-y: auto;"
    >
      <div
        style="text-align: center;"
      >
        <h4
          class="ant-typography"
          data-testid="text-welcome"
        >
          Welcome to Snoop-hub :)
        </h4>
      </div>
      <div
        style="display: flex; justify-content: center; margin: 16px 0px 40px;"
      />
    </div>
    <div
      class="Toastify"
    />
  </main>
</DocumentFragment>
`;

exports[`<MainPage /> search by inputted value 1`] = `
<DocumentFragment>
  <main
    class="container"
  >
    <div
      style="margin: 100px 0px 16px; padding: 0px 16px; width: 100%;"
    >
      <span
        class="ant-input-search ant-input-search-enter-button ant-input-group-wrapper"
      >
        <span
          class="ant-input-wrapper ant-input-group"
        >
          <input
            class="ant-input"
            data-testid="input-search"
            placeholder="Search github user"
            type="text"
            value="ardaplun"
          />
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn ant-input-search-button ant-btn-primary"
              type="button"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
      </span>
    </div>
    <div
      style="text-align: center;"
    >
      <h4
        class="ant-typography"
        data-testid="text-found"
      >
        Found 1 for ardaplun user
      </h4>
    </div>
    <div
      style="flex: 1; width: 100%; padding: 16px; overflow-y: auto;"
    >
      <div
        class="ant-row ant-row-space-around ant-row-middle"
      >
        <div
          class="ant-col ant-col-11"
          data-testid="result-search-ardaplun"
          style="border: 1px solid #ccc; cursor: pointer; display: flex; flex-direction: column; align-items: center; padding: 10px; margin-bottom: 16px;"
        >
          <img
            alt="Avatar of ardaplun"
            data-testid="result-search-ardaplun-img"
            height="60px"
            loading="lazy"
            src="https://avatars2.githubusercontent.com/u/17042686?v=4"
            width="60px"
          />
          <b>
            ardaplun
          </b>
        </div>
      </div>
      <div
        style="display: flex; justify-content: center; margin: 16px 0px 40px;"
      />
    </div>
    <div
      class="Toastify"
    />
  </main>
</DocumentFragment>
`;
